Polish Microarray GSEA example (#434)
AlexsLemonade/refinebio-examples: 02-microarray/pathway-analysis_microarray_02_gsea.Rmd

# Purpose of this analysis

This example is one of pathway analysis module set, we recommend looking at the [pathway analysis table below](#how-to-choose-a-pathway-analysis) to help you determine which pathway analysis method is best suited for your purposes.

This particular example analysis shows how you can use Gene Set Enrichment Analysis (GSEA) to detect situations where genes in a predefined gene set or pathway change in a coordinated way between two biological states [@Subramanian2005; @GSEA-broad-institute]. 
Changes at the pathway-level may be statistically significant, and contribute to phenotypic differences, even if the changes in the expression level of individual genes are small. 

> 

Author: Okay! I'll give rearranging a try.

> GSEA calculates a pathway-level metric, called an enrichment score (ES), by ranking genes by a gene-level statistic.

Author: Sure, but we absolutely will have to say NES because of what `GSEA()` returns...

> GSEA calculates a pathway-level metric, called an enrichment score (ES), by ranking genes by a gene-level statistic.

Reviewer: Yeah, in the context of a variable, it is unavoidable...

> This score reflects whether or not a gene set or pathway is overrepresented at the top or bottom of the gene rankings [@clusterProfiler-book]. 

Author: I just added 2020? I don't know what else we would use.

dge_df
```

## Getting familiar with MSigDB gene sets available via `msigdbr`

The Molecular Signatures Database (MSigDB) is a resource that contains annotated gene sets that can be used for pathway or gene set analyses [@Subramanian2005]. 
We can use the `msigdbr` package to access these gene sets in a format compatible with the package we'll use for analysis, `clusterProfiler` [@Igor2020; @Yu2012].

The gene sets available directly from MSigDB are applicable to human studies.
`msigdbr` also supports commonly studied model organisms.

Let's take a look at what organisms the package supports with `msigdbr_species()`.

```{r}
msigdbr_species()
```

> MSigDB contains [8 different gene set collections](https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp) [@Subramanian2005] that are distinguished by how they are derived (e.g., computationally mined, curated).

Reviewer: We should probably use the preferred citation here: (Liberzon et al. 2011)
```
@article{Liberzon2011,
  doi = {10.1093/bioinformatics/btr260},
  url = {https://doi.org/10.1093/bioinformatics/btr260},
  year = {2011},
  month = may,
  publisher = {Oxford University Press ({OUP})},
  volume = {27},
  number = {12},
  pages = {1739--1740},
  author = {Arthur Liberzon and Aravind Subramanian and Reid Pinchback and Helga Thorvaldsdóttir and Pablo Tamayo and Jill P Mesirov},
  title = {Molecular signatures database ({MSigDB}) 3.0},
  journal = {Bioinformatics}
}
```

Author: Ah, you mean in addition to `@Subramanian2005` correct?

> To cite your use of the Molecular Signatures Database (MSigDB), a joint project of UC San Diego and Broad Institute, please reference Subramanian, Tamayo, et al. (2005, PNAS 102, 15545-15550) and one or more of the following as appropriate: Liberzon, et al. (2011, Bionformatics), Liberzon, et al. (2015, Cell Systems), and also the source for the gene set as listed on the gene set page. 

Where the 2015 citation is hallmarks, specifically.

Reviewer: Yes, in addition...

Author: 👍  I propagated this throughout the pathway analysis notebooks I've gone through so far.

Let's create a data frame that shows the mapped Entrez IDs along with the differential expression stats for the respective Ensembl IDs.

```{r}

> # First let's create a mapped data frame we can join to the differential

Author: @jashapiro does this section need more info about `org.Dr.eg.db` per #435 or do you think it's sufficient to link to the more detailed identifier conversion example?

Reviewer: Maybe a little note about where the mapIds() function comes from? Also, in those examples I moved `keytype` to right after `key` for "my brain works this way" reasons.

Also, is there a reason `entrez_id` is in quotes? I don't think we do that elsewhere.

Author: > Maybe a little note about where the mapIds() function comes from? Also, in those examples I moved `keytype` to right after `key` for "my brain works this way" reasons.

👍 
> Also, is there a reason `entrez_id` is in quotes? I don't think we do that elsewhere.

Not that I'm aware of, I can change that.

### Run GSEA using the `GSEA()` function

Genes were ranked from most positive to most negative, weighted according to their gene-level statistic, in the previous section.
In this section, we will implement GSEA to calculate the enrichment score for each gene set using our pre-ranked gene list.

The GSEA algorithm utilizes random sampling so we are going to set the seed to make our results reproducible.

```{r}
# Set the seed so our results are reproducible:

> set.seed(2020)

Reviewer 2: > We set the seed in the RNA-seq example, but we didn't here. I added a set.seed() step here, right before we "need" it should someone be viewing this example and have jumped to the analysis code. Is this the correct thing to do?

Yes, this is correct. At the time of authoring this example, it was not obvious that the `set.seed()` step also needed to be added before the `GSEA()` function, which is why it was left out here. Good point re the placement of the step as well!

)
```

Significance is assessed by permuting the gene labels of the pre-ranked gene list and recomputing the enrichment scores of the gene set for the permuted data, which generates a null distribution for the enrichment score.

> The `pAdjustMethod` arguments to `GSEA()` above specifies what method to use for adjusting the p-values to account for multiple hypothesis testing; the `pvalueCutoff` argument tells the function to only return pathways with adjusted p-values less than that threshold in the `results` slot.

Reviewer: ```suggestion
The `pAdjustMethod` argument to `GSEA()` above specifies what method to use for adjusting the p-values to account for multiple hypothesis testing; the `pvalueCutoff` argument tells the function to only return pathways with adjusted p-values less than that threshold in the `results` slot.
```

### Most Positive NES

Let's look at the 3 gene sets with the most positive NES.

```{r rownames.print = FALSE}
gsea_result_df %>%

>   # This orders the results by the NES value and takes the top 3 rows
>   dplyr::slice_max(n = 3, order_by = NES)

Reviewer: In this spirit of replacing "top" with "max"...
```suggestion
  # This returns the 3 rows with the largest NES values
  dplyr::slice_max(n = 3, order_by = NES)
```

Author: Fair enough!

### Most Negative NES

Let's look for the 3 gene sets with the most negative NES.

```{r rownames.print=FALSE}
gsea_result_df %>%

>   # This orders the results by the NES value and takes the bottom 3 rows

Reviewer: ```suggestion
  # Return the 3 rows with the smallest (most negative) NES values
```
